File: Hard/the_skyline_problem/skyline_problem.cpp
#include <vector>
#include <queue>
#include <algorithm>
#include <iostream>
#include <set>
using namespace std;

class Solution {
public:
    vector<vector<int>> getSkyline(vector<vector<int>>& buildings) {
        vector<pair<int,int>> events; // (x, height)
        int n = buildings.size();

        // Use index-based iteration instead of 'auto &b'
        for (int i = 0; i < n; i++) {
            int L = buildings[i][0];
            int R = buildings[i][1];
            int H = buildings[i][2];

            events.push_back({L, -H}); // start event, negative height
            events.push_back({R, H});  // end event, positive height
        }

        // Sort by x; start (-height) before end (+height) if tied
        sort(events.begin(), events.end());

        multiset<int> heights;
        heights.insert(0); // ground level

        vector<vector<int>> result;
        int prev_max = 0;

        for (int i = 0; i < events.size(); i++) {
            int x = events[i].first;
            int h = events[i].second;

            if (h < 0) {
                // Start of building
                heights.insert(-h);
            } else {
                // End of building
                heights.erase(heights.find(h));
            }

            int current_max = *heights.rbegin();
            if (current_max != prev_max) {
                result.push_back({x, current_max});
                prev_max = current_max;
            }
        }

        return result;
    }
};

// Example usage:
int main(){
    Solution sol;
    vector<vector<int>> buildings1 = {{2,9,10},{3,7,15},{5,12,12},{15,20,10},{19,24,8}};
    vector<vector<int>> skyline1 = sol.getSkyline(buildings1);

    cout << "Skyline for buildings1: {2,9,10},{3,7,15},{5,12,12},{15,20,10},{19,24,8}: ";
    for (int i = 0; i < skyline1.size(); i++) {
        cout << "[" << skyline1[i][0] << "," << skyline1[i][1] << "]";
        if (i != skyline1.size() - 1){
            cout << ", ";
        }
    }
    cout << endl;

    vector<vector<int>> buildings2 = {{0,2,3},{2,5,3}};
    vector<vector<int>> skyline2 = sol.getSkyline(buildings2);
    cout << "Skyline for buildings2: {0,2,3},{2,5,3} ";
    for (int i = 0; i < skyline2.size(); i++) {
        cout << "[" << skyline2[i][0] << "," << skyline2[i][1] << "]";
        if (i != skyline2.size() - 1){
            cout << ", ";
        }
    }
    cout << endl;
    return 0;
}

File: Hard/the_skyline_problem/skyline_problem.py
from typing import List
import heapq

def getSkyline(buildings: List[List[int]]) -> List[List[int]]:
    events = []
    for L, R, H in buildings:
        events.append((L, -H, R))  # start of building
        events.append((R, 0, 0))   # end of building

    # Sort after all events are added
    events.sort()

    result = []
    heap = [(0, float('inf'))]  # (-height, end_x)
    prev_max = 0

    for x, negH, R in events:
        # Remove buildings that ended
        while heap[0][1] <= x:
            heapq.heappop(heap)

        # If start event, add to heap
        if negH != 0:
            heapq.heappush(heap, (negH, R))

        # Current max height
        current_max = -heap[0][0]

        if current_max != prev_max:
            result.append([x, current_max])
            prev_max = current_max

    return result
    
if __name__ == "__main__":
    print("Testing skyline function:")
    buildings = [[2,9,10],[3,7,15],[5,12,12],[15,20,10],[19,24,8]]
    print("For the input buildings =", buildings, ", the skyline is:", getSkyline(buildings))  # Expected output: [[2,10],[3,15],[7,12],[12,0],[15,10],[20,8],[24,0]]

    buildings2 = [[0,2,3],[2,5,3]]
    print("For the input buildings =", buildings2, ", the skyline is:", getSkyline(buildings2))  # Expected output: [[0,3],[5,0]]

